### PR TITLE
Equipment Shop: Use ORM and migrations instead of custom sqlcode

### PR DESCRIPTION
**Pull request by Histalek**
This PR is an implementation example of the ORM and migration libraries for discussion and test purposes.

Will be rebased on master once #594 and #673 are merged.

### CODE REVIEW

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 1]
```diff
@@ -27,8 +27,21 @@ local serverColsVar = GetConVar("ttt_bem_sv_cols")
...
--- add favorites DB functions
-include("favorites_db.lua")
+-- setup database
+if not sql.TableExists("ttt_bem_fav") then
+	sql.Query("CREATE TABLE ttt_bem_fav (guid TEXT, role TEXT, weapon_id TEXT)")
+end
+
+local upQueryV1 = "CREATE TABLE ttt_equip_fav (guid TEXT, role TEXT, weapon_id TEXT, PRIMARY KEY (guid, role, weapon_id)); " ..
+					"INSERT INTO ttt_equip_fav SELECT * FROM ttt_bem_fav; " ..
+					"DROP TABLE ttt_bem_fav;"
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:37]: I'd remove everything that has something todo with `bem`. Just rename it to `shop`

Comment (Histalek) [reply gamemodes/terrortown/gamemode/client/cl_equip.lua:37]: that is out of scope for this PR. Also this part is needed as we don't want to throw away already existing data.

What could be discussed is making this migration irreversible by not specifying a "down" query (irreversible migrations will still need to get handled in the migration library and also documented thoroughly)

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 2]
```diff
@@ -27,8 +27,21 @@ local serverColsVar = GetConVar("ttt_bem_sv_cols")
 local serverRowsVar = GetConVar("ttt_bem_sv_rows")
 local serverSizeVar = GetConVar("ttt_bem_sv_size")
 
--- add favorites DB functions
-include("favorites_db.lua")
+-- setup database
+if not sql.TableExists("ttt_bem_fav") then
+	sql.Query("CREATE TABLE ttt_bem_fav (guid TEXT, role TEXT, weapon_id TEXT)")
+end
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:33]: this part needs to be replaced by a migration as well, but as this table was already created on many servers without the migration library I first need to implement something in the migration library to handle these kinds of situations

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 3]
```diff
@@ -27,8 +27,21 @@ local serverColsVar = GetConVar("ttt_bem_sv_cols")
...
+
+local upQueryV1 = "CREATE TABLE ttt_equip_fav (guid TEXT, role TEXT, weapon_id TEXT, PRIMARY KEY (guid, role, weapon_id)); " ..
+					"INSERT INTO ttt_equip_fav SELECT * FROM ttt_bem_fav; " ..
+					"DROP TABLE ttt_bem_fav;"
+local downQueryV1 = "CREATE TABLE ttt_bem_fav (guid TEXT, role TEXT, weapon_id TEXT); " ..
+					"INSERT INTO ttt_bem_fav SELECT * FROM ttt_equip_fav; " ..
+					"DROP TABLE ttt_equip_fav;"
+
+migration.Add("equipment_Favorites", 1, upQueryV1, downQueryV1)
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:42]: The `upQueryV1` is a sql query which describes the changes to be made to the database schema.
The `downQueryV1` is the sql query neccessary to reverse these changes.

The first and seconds argument to `migration.Add()` are used to give more explizit control over the schema changes (used in other migration functions)

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 4]
```diff
@@ -27,8 +27,21 @@ local serverColsVar = GetConVar("ttt_bem_sv_cols")
...
+
+local equipFavModel = orm.Make("ttt_equip_fav")
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:44]: this creates a model of the specified databasetable using the orm library

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 5]
```diff
@@ -340,9 +353,7 @@ local function CreateEquipmentList(t)
 				-- Favorites marker icon
 				ic.favorite = false
 
-				local favorites = GetFavorites(steamid, currole)
-
-				if favorites and IsFavorite(favorites, item.id) then
+				if equipFavModel:Find({steamid, currole, item.id}) then
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:356]: check if an object of the model gets created if it does it exists in the database

File: gamemodes/terrortown/gamemode/client/cl_equip.lua [hunk 6]
```diff
@@ -832,12 +843,19 @@ function TraitorMenuPopup()
...
 		local steamid = ply:SteamID64()
 
-		CreateFavTable()
-
 		if pnl.favorite then
-			RemoveFavorite(steamid, role, weapon)
+			local favorite = equipFavModel:Find({steamid, role, weapon})
+			favorite:Delete()
+			-- possible one-liner alternative:
+			--equipFavModel:Find({steamid, role, weapon}):Delete()
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_equip.lua:850]: creates an object of the model with values from the database and then deletes it (maybe needs an `IsValid` check just to be safe)

**Review by Histalek — COMMENTED** [review]
i commented on all parts of the implementation example to better highlight them

**Pull request closed without merging**